From f1e98581dad2f1d72106034a73f05f60dcee4de1 Mon Sep 17 00:00:00 2001
From: Josh Stone <jistone@redhat.com>
Date: Wed, 29 Jul 2020 11:50:36 -0700
Subject: [PATCH 1005/1013] [rust] Skip processor info on i386 mingw-w64 < 7

This will limit LLVM thread pools to a single thread on that target, but
we don't use that functionality in rustc anyway.

See also:
https://github.com/llvm/llvm-project/commit/8404aeb56a73ab24f9b295111de3b37a37f0b841#commitcomment-37406150
---
 llvm/lib/Support/Windows/Threading.inc | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/llvm/lib/Support/Windows/Threading.inc b/llvm/lib/Support/Windows/Threading.inc
index 7b48ca8fb1fb..2911485d4a29 100644
--- a/llvm/lib/Support/Windows/Threading.inc
+++ b/llvm/lib/Support/Windows/Threading.inc
@@ -142,6 +142,11 @@ struct ProcessorGroup {
 
 template <typename F>
 static bool IterateProcInfo(LOGICAL_PROCESSOR_RELATIONSHIP Relationship, F Fn) {
+#if !defined(_WIN64) && defined(__MINGW64_VERSION_MAJOR) && __MINGW64_VERSION_MAJOR < 7
+  // `GetLogicalProcessorInformationEx@12` was only added to i386 mingw-w64 in v7.0.0
+  // https://github.com/mingw-w64/mingw-w64/commit/24842d45e025db0d38fa2bbd932b95a83282efa2#diff-faf1d8a1556e75a84b7cef2e89512e79R634
+  return false;
+#else
   DWORD Len = 0;
   BOOL R = ::GetLogicalProcessorInformationEx(Relationship, NULL, &Len);
   if (R || GetLastError() != ERROR_INSUFFICIENT_BUFFER) {
@@ -162,6 +167,7 @@ static bool IterateProcInfo(LOGICAL_PROCESSOR_RELATIONSHIP Relationship, F Fn) {
   }
   free(Info);
   return true;
+#endif
 }
 
 static ArrayRef<ProcessorGroup> getProcessorGroups() {
-- 
2.35.1

